src/main/java/org/kizombadev/markdownparser/SyntaxAnalyzer.java

```diff
@@ -0,0 +1,121 @@
+package org.kizombadev.markdownparser;
+
+import com.google.common.collect.ImmutableList;
+import org.kizombadev.markdownparser.entities.*;
+
+import java.util.Stack;
+
+public class SyntaxAnalyzer {
+
+    private ImmutableList<Token> tokens;
+    private int tokenIndex = 0;
+    private Stack<Token.Category> stack = new Stack<>();
+
+    public static SyntaxAnalyzer create() {
+        return new SyntaxAnalyzer();
+    }
+
+    public Syntax parse(ImmutableList<Token> tokens) {
+        this.tokens = tokens;
+
+        RootSyntax root = new RootSyntax();
+
+        while (currentToken() != null) {
+            if (currentToken().getCategory().equals(Token.Category.NUMBER_SIGN)) {
+                handleBigHeadlineLine(root);
+            } else if (currentToken().getCategory().equals(Token.Category.DOUBLE_NUMBER_SIGN)) {
+                handleSmallHeadlineLine(root);
+            } else if (currentToken().getCategory().equals(Token.Category.GREATER_THAN_SIGN)) {
+                handleQuotation(root);
+            }
+
+            if (currentToken() != null && currentToken().getCategory().equals(Token.Category.NEW_LINE)) {
+                stepTokenForward();
+            }
+
+
+        }
+
+        return root;
+    }
+
+    private void handleQuotation(RootSyntax currentRoot) {
+        Syntax quotation = new Quotation();
+        stepTokenForward();
+
+        handleLineContainer(quotation);
+        currentRoot.addChild(quotation);
+    }
+
+    private void handleBigHeadlineLine(Syntax currentRoot) {
+        Syntax bigHeadline = new BigHeadline();
+        stepTokenForward();
+
+        handleLineContainer(bigHeadline);
+        currentRoot.addChild(bigHeadline);
+    }
+
+    private void handleSmallHeadlineLine(Syntax currentRoot) {
+        Syntax smallHeadline = new SmallHeadline();
+        stepTokenForward();
+
+
+        handleLineContainer(smallHeadline);
+        currentRoot.addChild(smallHeadline);
+    }
+
+    private void handleLineContainer(Syntax currentRoot) {
+
+        while (currentToken() != null && !currentToken().getCategory().equals(Token.Category.NEW_LINE)) {
+
+            if (currentToken() != null && currentToken().getCategory().equals(Token.Category.DOUBLE_STAR)) {
+                handleBold(currentRoot);
+            } else if (currentToken() != null && currentToken().getCategory().equals(Token.Category.TEXT)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'currentToken' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS2259)